tools/tests/test_portability.py

```
@@ -0,0 +1,280 @@
+    charter = (bracketed / "charter" / "CHARTER.md").read_text(encoding="utf-8")
+
+    for command in commands:
+        # Claude Code's half of the contract: it substitutes the placeholder
+        # into the command string before any shell sees it. Codex performs no
+        # substitution at all -- it supplies the same value through the
+        # environment, which `_hook_env` sets. Modelling both is what lets a
+        # lawful emitter of either shape pass. `check_delivery`'s
+        # PLUGIN_ROOT_REF makes the same substitution assumption, so the two
+        # guards drift together or not at all.
+        expanded = command.replace("${CLAUDE_PLUGIN_ROOT}", bracketed.as_posix())
+        result = subprocess.run(
```

Ruff (0.16.2)

[error] `subprocess` call with `shell=True` identified, security issue

(S602)